backup/moodle2/restore_cms_stepslib.php

```diff
@@ -0,0 +1,164 @@
+    protected $afterparty = [];
+
+    /**
+     * Add a datasource restore processor object to the afterparty.
+     *
+     * @param object $obj
+     */
+    public function add_to_after_party($obj) {
+        $this->afterparty[] = $obj;
+    }
+
+    /**
+     * Define the structure for the restore
+     */
+    protected function define_structure() {
+        $paths = [];
+        $userinfo = $this->get_setting_value('userinfo');
+
+        $paths[] = new restore_path_element('cms', '/activity/cms');
+        $paths[] = new restore_path_element('cms_types', '/activity/cms/cms_types');
+
+        foreach (dsbase::get_datasources(null, false) as $ds) {
+            $paths = $ds->restore_define_structure($paths, $this);
+        }
+
+        // Return the paths wrapped into standard activity structure.
+        return $this->prepare_activity_structure($paths);
+    }
+
+    /**
+     * Restore the cms data.
+     *
+     * @param array $data
+     */
+    protected function process_cms(array $data) {
+        global $DB;
+
+        $data = (object)$data;
+        $data->course = $this->get_courseid();
+        $data->typeid = 0;
+
+        // Insert the record.
+        $newitemid = $DB->insert_record('cms', $data);
+
+        // Immediately after inserting "activity" record, call this.
+        $this->apply_activity_instance($newitemid);
+    }
+
+    /**
+     * Restore the cms type.
+     *
+     * @param array $data
+     */
+    protected function process_cms_types(array $data) {
+        global $DB;
+
+        $data = (object)$data;
+        $oldid = $data->id;
+
+        $typeid = $this->find_existing_cms_type($data);
+
+        if ($typeid === false) {
+            $typeid = $DB->insert_record('cms_types', $data);
+        }
+
+        $this->set_mapping('cms_types', $oldid, $typeid, true);
+
+        $DB->update_record('cms', (object) ['id' => $this->get_new_parentid('cms'), 'typeid' => $typeid]);
+    }
+
+    /**
+     * Looks for an existing record that matches the CMS type.
+     *
+     * @param stdClass $data
+     * @return int/false The db ID of the CMS type or false if none was found.
+     */
+    protected function find_existing_cms_type(\stdClass $data) {
+        global $DB;
+
+        if ($typeid = $this->get_mappingid('cms_types', $data->id)) {
+            return $typeid;
+        }
+
+        if ($this->task->is_samesite()) {
+            if ($DB->record_exists('cms_types', ['id' => $data->id])) {
+                return $data->id;
+            }
+        }
+
+        // Try to find a CMS type is that is a match for content. Returns false if none found.
+        $sql = "SELECT id FROM {cms_types} WHERE " .
+                "name = :name AND idnumber = :idnumber AND datasources = :datasources AND " .
+                $DB->sql_compare_text('title_mustache') . ' = ' . $DB->sql_compare_text(':title_mustache') . " AND " .
+                $DB->sql_compare_text('mustache') . ' = ' . $DB->sql_compare_text(':mustache');
```

Comment:
```suggestion
        $sql = "SELECT id
                       FROM {cms_types}
                     WHERE name = :name
                           AND idnumber = :idnumber
                           AND datasources = :datasources
                           AND $DB->sql_compare_text('title_mustache') = $DB->sql_compare_text(':title_mustache')
                           AND $DB->sql_compare_text('mustache') = $DB->sql_compare_text(':mustache')";
```
based on https://moodledev.io/general/development/policies/codingstyle/sql#indentation

Reply:
done